src/controlobjectscript.cpp

```diff
@@ -0,0 +1,55 @@
+#include <QApplication>
+#include <QtDebug>
+
+#include "controlobjectscript.h"
+
+ControlObjectScript::ControlObjectScript(const ConfigKey& key, QObject* pParent)
+        : ControlObjectSlave(key, pParent) {
+}
+
+void ControlObjectScript::connectScriptFunction(
+        const ControllerEngineConnection& conn) {
+    if (m_connectedScriptFunctions.isEmpty()) {
+        // we connect the slots only, if there will be actually a script
+        // connected
+        connect(m_pControl.data(), SIGNAL(valueChanged(double, QObject*)),
+                this, SLOT(slotValueChanged(double,QObject*)),
+                Qt::QueuedConnection);
+        connect(this, SIGNAL(trigger(double, QObject*)),
+                this, SLOT(slotValueChanged(double,QObject*)),
+                Qt::QueuedConnection);
+    }
+    m_connectedScriptFunctions.append(conn);
+}
+
+bool ControlObjectScript::disconnectScriptFunction(
+        const ControllerEngineConnection& conn) {
+    bool ret = m_connectedScriptFunctions.removeAll(conn) > 0;
+    if (m_connectedScriptFunctions.isEmpty()) {
+        // no script left, we can disconnected
+        disconnect(m_pControl.data(), SIGNAL(valueChanged(double, QObject*)),
+                this, SLOT(slotValueChanged(double,QObject*)));
+        disconnect(this, SIGNAL(trigger(double, QObject*)),
+                this, SLOT(slotValueChanged(double,QObject*)));
+    }
+    return ret;
+}
+
+void ControlObjectScript::slotValueChanged(double value, QObject*) {
+    // Make a local copy of m_connectedScriptFunctions fist.
+    // This allows a script to disconnect a callback from the callback
+    // itself. Otherwise the this may crash since the disconnect call
+    // happens during conn.function.call() in the middle of the loop below.
+    QList<ControllerEngineConnection> connections = m_connectedScriptFunctions;
+    for(auto&& conn: connections) {
```

Comment:
> (It would also be nice to know what Qt does that the STL containers don't that allows auto& to work (and whether a copy occurs in that case)... )

auto& works for std::vector<int> but not for std::vector<bool>. 
http://en.cppreference.com/w/cpp/container/vector_bool

This is because of std::vector<bool> operator[] and the definition of std::vector<bool>::reference. 
Instead of a reference directly to the storage location an rvalue reference to a proxy object is returned. It features an operator=, which makes my example above working. 
The Qt containers do not allow to return proxy objects instead of direct lvalue references. 

Conclusion: 
"auto&" should be save for the Qt case and in code where the container is known. 
In template implementations were this is not true "auto&&" is the way to go. 

I have no preferences if we should use always "auto&&" for the non const case in Mixxx
* "auto&" should rise less questions for new contributors
* "auto&&" will always work and more easy once you know it.  

So should I revert https://github.com/daschuer/mixxx/commit/1618b76753c3ca1fc2d6c56d8b963cb28e226cf8
In any case we should add a not to our style guides.